---
title: "DESEq2: Arabidopsis's root RNA-seq under salt treatments"
output: html_notebook
---
1) to check baseline of SUMOylation gene expression in Arabidopsis roots
2) to perform DE analysis on salt treated vs control samples

# load libraries
```{r}
library(tidyverse)
library(data.table)
library(DESeq2)
library(limma)
library(ComplexHeatmap)
library(circlize)
```

# import data
```{r}
# list all the files
file_list <- list.files(path="../Raw_FPKM_TPM/",
                        pattern=".txt",
                        full.names=TRUE)


file_list_short <- lapply(file_list, function(x) {
  # Step 1: Split the file path by "/"
  path_parts <- strsplit(x, split = "/")[[1]]
  
  # Step 2: Get the filename (the last part of the path)
  file_name <- path_parts[length(path_parts)]
  
  # Step 3: Split the filename by "_" and extract the first part
  file_name_parts <- strsplit(file_name, split = "_")[[1]]
  file_name_parts[1]
})

file_data <- lapply(file_list, fread)

# Combine them into one table (fill missing columns with NA)
merged_data <- rbindlist(file_data, use.names = TRUE, fill = TRUE, idcol = TRUE)

# to check which file has duplicate gene ids
dup_file <- merged_data %>% group_by(Gene.ID, .id) %>%
  dplyr::summarise(n = dplyr::n(), .groups = "drop") %>%
  dplyr::filter(n > 1L) # this gene: AT1G58520
#unique(dup_file$Gene.ID)

# remove the row of the duplicates and make it wide format
merged_data_raw_wideF <- merged_data %>%
  group_by(.id, Gene.ID) %>% 
  filter(!(Gene.ID == "AT1G58520" &Raw > 0 & TPM == 0)) %>% #this gene has duplicate rows with FPKM&TPM=0 but raw >0, so keep only the one with FPKM&TPM&raw >0
  select(".id","Gene.ID", "Raw") %>% 
  ungroup() %>% 
  pivot_wider(names_from = ".id", values_from = "Raw") %>% 
  column_to_rownames(var = "Gene.ID")

merged_data_TPM_wideF <- merged_data %>%
  group_by(.id, Gene.ID) %>% 
  filter(!(Gene.ID == "AT1G58520" &Raw > 0 & TPM == 0)) %>% #this gene has duplicate rows with FPKM&TPM=0 but raw >0, so keep only the one with FPKM&TPM&raw >0
  select(".id","Gene.ID", "TPM") %>% 
  ungroup() %>% 
  pivot_wider(names_from = ".id", values_from = "TPM") %>% 
  column_to_rownames(var = "Gene.ID")


# add column names
for (i in 1:68) {
  colnames(merged_data_raw_wideF)[i] <- file_list_short[[i]]
  colnames(merged_data_TPM_wideF)[i] <- file_list_short[[i]]
}


# import metadata 
metadata <- read.csv("../Arabidopsis_salt_root_RNAseq_meta.csv")
only_salt_meta <- metadata %>% filter(Salt.only == "Y") #46

# only salt
merged_data_raw_wideF_salt <- merged_data_raw_wideF %>% select(only_salt_meta$SRA)

# SUMOylation gene list
SUMO_genes <- read.csv("SUMOylation_AT_genes_sepProtease.csv")
```

# Plot functions
```{r}
#plotPCA
# GET_PCA
get_PCA <- function(input_data = NULL, 
                    input_meta = NULL, 
                    colour_palette = NULL, 
                    name = NULL) {
  
  require(FactoMineR)
  require(factoextra)
  require(ggplot2)
  pca <- PCA(t(input_data), graph = FALSE)
  p <- fviz_pca_ind(pca,
             geom.ind = "point", 
             col.ind = input_meta, # color by groups
             palette = colour_palette,
             addEllipses = TRUE, # Concentration ellipses
             legend.title = "Groups"
             )
  
  ggsave(paste(name,"_PCA.png", sep = ""),
       plot = p,
       scale = 1,
       width = 7,
       height = 7,
       units = c("in"),
       dpi = 300)
}

```
```{r}
colnames(dataset_raw_dis0) == only_salt_meta$SRA
```


# Quality controls
```{r}
# prepare data for PCA
# discard all 0
dataset_raw_dis0 <- merged_data_raw_wideF_salt %>% filter_if(is.numeric, any_vars(.!=0)) 

# just for the PCA
only_salt_meta_edit <- only_salt_meta[c(1:40,43:44),]
  
dds <- DESeqDataSetFromMatrix(
  countData = dataset_raw_dis0, # raw counts
  colData = only_salt_meta_edit, 
  design = ~condition# Here we are specifying a model by group 
)

dds_norm <- vst(dds) # blind is TRUE, the design doesn't affect the normalization
VST_normCount <- as.data.frame(assay(dds_norm))

res.pca_1 <- prcomp(t(VST_normCount), scale = T)
res.pca_1_df <- as.data.frame(res.pca_1$x)
res.pca_1_df <- res.pca_1_df %>% rownames_to_column() %>% left_join(only_salt_meta_edit, by = c("rowname"="SRA")) %>% column_to_rownames()

percentage_1 <- round((res.pca_1$sdev)^2 / sum((res.pca_1$sdev)^2) * 100, 2)
percentage_1 <- paste(colnames(res.pca_1_df)," ","(", paste(as.character(percentage_1), "%", ")", sep=""), sep = "")


# plot PCA - colour by experiment (BioProject)

ggplot(res.pca_1_df,aes(x=PC1,y=PC2,color=Project )) +
  geom_point()+theme_bw() + xlab(percentage_1[1]) + ylab(percentage_1[2]) +
  labs(title  = "Bulk RNA-seq: root & salt (61), colored by Bioproject",
       subtitle = "1) discard exp == 0 (30,867 genes), scaled")


# plot PCA - colour by treatments
ggplot(res.pca_1_df,aes(x=PC1,y=PC2,color=condition )) +
  geom_point()+theme_bw() + xlab(percentage_1[1]) + ylab(percentage_1[2]) +
  labs(title  = "Bulk RNA-seq: root & salt (61), colored by condition",
       subtitle = "1) discard exp == 0 (30,867 genes), scaled")
# Do we need batch correction? -YES

# Perform batch correction
only_salt_meta_edit <- only_salt_meta_edit %>% 
  mutate(condition = str_replace(condition, "NaCl", "Salt")) %>% 
  mutate(condition = str_replace(condition, "control", "Ctrl")) %>% 
  mutate(Project2 = case_when(Project == "GSE275862" ~ "1",
                          Project == "GSE252395" ~ "2",
                          Project == "GSE270516" ~ "3",
                          Project == "GSE229217" ~ "4",
                          Project == "GSE124866" ~ "5",
                          str_detect(Project,"PRJNA904764") ~ "6, 8, 9",
                          Project == "SUMOcode" ~ "7"
                          ))

VST_normCount_corrected <- as.data.frame(removeBatchEffect(x = VST_normCount, batch = only_salt_meta_edit$Project2))

get_PCA(input_data = VST_normCount, input_meta = only_salt_meta_edit$condition,colour_palette = c("blue", "red"), name = "colour_by_treat_beforeBatch")

#get_PCA(input_data = VST_normCount, input_meta = only_salt_meta_edit$Project2,colour_palette = c("#00AFBB", "#E7B800", "#FC4E07", "pink", "green", "blue", "black", "red"), name = "colour_by_proj2_beforeBatch")

png("colour_by_proj2_beforeBatch.png", width = 7, height = 7, res = 300, units = "in")
fviz_pca_ind(res.pca_1, habillage = only_salt_meta_edit$Project2, geom.ind = "point", addEllipses = T )
dev.off()

get_PCA(input_data = VST_normCount_corrected, input_meta = only_salt_meta_edit$condition,colour_palette = c("blue", "red"), name = "colour_by_treat_afterBatchProject")

#get_PCA(input_data = VST_normCount_corrected, input_meta = only_salt_meta_edit$Project2,colour_palette = c("#00AFBB", "#E7B800", "#FC4E07", "pink", "green", "blue", "black", "red"), name = "colour_by_proj_afterBatch")
res.pca_2 <- prcomp(t(VST_normCount_corrected), scale = T)

png("colour_by_proj2_afterBatch.png", width = 7, height = 7, res = 300, units = "in")
fviz_pca_ind(res.pca_2, habillage = only_salt_meta_edit$Project2, geom.ind = "point", addEllipses = T )
dev.off()

write.csv(VST_normCount_corrected, "Arabidopsis_root_salt_VST_normCount_corrected.csv")
write.csv(only_salt_meta_edit, "Arabidopsis_root_salt_metadata.csv")
```

# Baseline of SUMOylation genes
```{r}
# Use TPM
dataset_TPM_SUMO <- merged_data_TPM_wideF %>% rownames_to_column(var = "Gene.ID") %>% filter(Gene.ID %in% SUMO_genes$TAIR.ID) %>% left_join(SUMO_genes, by = c("Gene.ID"="TAIR.ID"))

# Heatmap
dataset_TPM_z <- as.data.frame(t(apply(dataset_TPM_SUMO[,-c(1,63:65)], 1, function(x) (x - mean(x)) / sd(x))))
dataset_TPM_z_NA0 <- dataset_TPM_z %>% mutate_if(is.numeric, ~replace(., is.na(.), 0))


ha = rowAnnotation(gene_name = anno_text(SUMO_genes$Symbol, location = 0.5, just = "center",
    gp = gpar(fill = c(rep(1,8),rep(2,4),3,rep(4,2),rep(5,2),rep(6,8),rep(7,8)), col = "white"),
    width = max_text_width(SUMO_genes$Symbol)*1.2))

ha_col <- HeatmapAnnotation(condition = metadata2$condition,
                            project = metadata2$Project,
                            col = list(condition = c(control = "lightblue",
                                                     NaCl = "red",
                                                     mannitol = "orange",
                                                     PEG = "pink"),
                                      project = c(GSE275862 = "grey15",
                                                  GSE252395 = "grey30",
                                                  GSE270516 = "grey60",
                                                  GSE229217 = "grey80",
                                                  GSE124866 = "black")))
col_rnorm = colorRamp2(c(-3, 0, 3), c("blue", "white", "red"))

png("z-row_NA0_data-name.png", height = 22, width = 20, res = 300, units = "in")
Heatmap(dataset_TPM_z_NA0, 
        name = "Z_score_TPM" ,
        col = col_rnorm,
        column_labels = metadata2$Sample,
        top_annotation = ha_col,
        left_annotation = ha,
        cluster_rows = F, 
        row_split = factor(c(rep("SUMO",8),rep("E1",4),"E2",rep("E3",2),rep("E4",2),rep("ULP",8),rep("DeSI",8)), levels = c("SUMO","E1","E2","E3","E4","ULP","DeSI")), 
        row_gap = unit(3, "mm")) 
dev.off()

dataset_TPM_SUMO_lf <- dataset_TPM_SUMO %>% gather(., key = "SRA", value = "TPM", -Type, -Type2, -Symbol, -Gene.ID) %>% 
  left_join(., metadata2)

png("boxplot.png", , height = 12, width = 7, res = 300, units = "in")
ggplot(dataset_TPM_SUMO_lf, aes(x=Symbol, y = TPM, colour = condition)) +
  geom_boxplot(outliers = F) +
  facet_grid(rows = vars(Type2), scales = "free", space = "free") + 
  coord_flip() +
  theme_bw()
dev.off()
```


```{r}
# Use normalised data by DEseq2 
```

```{r}
meta_data_sub <- only_salt_meta %>% filter(Project == experiment[7])
  data_sub <-  merged_data_raw_wideF %>% select(meta_data_sub$SRA)
  data_sub_dis0 <- data_sub[rowSums(data_sub>0)>0,]

 dds <- DESeqDataSetFromMatrix(
  countData = data_sub, # raw counts
  colData = meta_data_sub, 
  design = ~ condition # specifying a model by group 
)
  dds <- DESeq(dds) 
  res <- as.data.frame(results(dds))
  res %>% filter(log2FoldChange >1, padj <0.05) %>% nrow()
  resultsNames(dds)
  plotCounts(dds, gene="AT2G32950", intgroup="condition")
  
  data_sub %>% rownames_to_column() %>% filter(rowname == "AT2G32950")
  res %>% rownames_to_column() %>% filter(rowname == "AT2G32950")
```

# DE analysis - separately becuase the samples are too different to merge
```{r}
#  function to analyse DEG
#function
get_DEG_sig <- function(input_data = NULL, input_meta = NULL) {
  require(tidyverse)
  require(DESeq2)
  
  dds <- DESeqDataSetFromMatrix(
  countData = input_data, # raw counts
  colData = input_meta, 
  design = ~ condition # specifying a model by group 
)
  dds <- DESeq(dds) 
  res <- as.data.frame(results(dds)) 
  res_sig <- res %>% filter(abs(log2FoldChange) >= 1, padj < 0.05)
  res_sig
}

get_foldChange_all <- function(input_data = NULL, input_meta = NULL) {
  require(tidyverse)
  require(DESeq2)
  
  dds <- DESeqDataSetFromMatrix(
  countData = input_data, # raw counts
  colData = input_meta, 
  design = ~condition # specifying a model by group 
)
  dds <- DESeq(dds) 
  res <- as.data.frame(results(dds)) 
  res
}
```


```{r}
# DEG

experiment <- unique(only_salt_meta$Project)
DEG_list <- list() # an empty list to store the results

for (i in 1:length(experiment)) {
  meta_data_sub <- only_salt_meta %>% filter(Project == experiment[i])
  data_sub <-  merged_data_raw_wideF %>% select(meta_data_sub$SRA)
  data_sub_dis0 <- data_sub[rowSums(data_sub>0)>0,]
  DEG_list[[experiment[i]]] <- get_foldChange_all(input_data = data_sub_dis0, input_meta = meta_data_sub)
  
}

DEG_list

# extract the DEG names
#DEG_gene_name_all <- only significant
library(UpSetR)
DEG_gene_name_up <- list(exp1 = rownames(DEG_list[[1]] %>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp2 = rownames(DEG_list[[2]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp3 = rownames(DEG_list[[3]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp4 = rownames(DEG_list[[4]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp5 = rownames(DEG_list[[5]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp6 = rownames(DEG_list[[6]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp7 = rownames(DEG_list[[7]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp8 = rownames(DEG_list[[8]]%>% filter(log2FoldChange>= 1, padj < 0.05)),
                      exp9 = rownames(DEG_list[[9]]%>% filter(log2FoldChange>= 1, padj < 0.05)))
png("DEG_up-regulted_upset.png", height = 5, width = 7, units = "in", res = 300)
upset(fromList(DEG_gene_name_up),nsets =9, order.by = "freq", nintersects = 20)
dev.off()


DEG_gene_name_down <- list(exp1 = rownames(DEG_list[[1]] %>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp2 = rownames(DEG_list[[2]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp3 = rownames(DEG_list[[3]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp4 = rownames(DEG_list[[4]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp5 = rownames(DEG_list[[5]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp6 = rownames(DEG_list[[6]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp7 = rownames(DEG_list[[7]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp8 = rownames(DEG_list[[8]]%>% filter(log2FoldChange<= (-1), padj < 0.05)),
                      exp9 = rownames(DEG_list[[9]]%>% filter(log2FoldChange<= (-1), padj < 0.05)))
png("DEG_down-regulted_upset.png", height = 5, width = 7, units = "in", res = 300)
upset(fromList(DEG_gene_name_down),nsets =9, order.by = "freq", nintersects = 20)
dev.off()


saveRDS(DEG_gene_name_up, "DEG_gene_name_up_Arabidopsis_9exp.rds")
saveRDS(DEG_gene_name_down, "DEG_gene_name_down_Arabidopsis_9exp.rds")

```

```{r}
#Fc_list

SUMO_foldChange_1 <- DEG_list[[1]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "1") 
SUMO_foldChange_2 <- DEG_list[[2]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "2") 
SUMO_foldChange_3 <- DEG_list[[3]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "3") 
SUMO_foldChange_4 <- DEG_list[[4]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "4") 
SUMO_foldChange_5 <- DEG_list[[5]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "5") 
SUMO_foldChange_6 <- DEG_list[[6]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "6") 
SUMO_foldChange_7 <- DEG_list[[7]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "7") 
SUMO_foldChange_8 <- DEG_list[[8]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "8") 
SUMO_foldChange_9 <- DEG_list[[9]] %>% rownames_to_column() %>% filter(rowname %in% SUMO_genes$TAIR.ID) %>% mutate(exp = "9") 


SUMO_foldChange_all <- rbind(SUMO_foldChange_1, SUMO_foldChange_2, SUMO_foldChange_3,SUMO_foldChange_4, SUMO_foldChange_5, SUMO_foldChange_6, SUMO_foldChange_7,SUMO_foldChange_8, SUMO_foldChange_9) %>% 
  right_join(., SUMO_genes, by = c("rowname"="TAIR.ID")) %>% 
  mutate(sig = ifelse(padj < 0.05, "sig", "no_sig")) 
  


SUMO_foldChange_all$Type2 <- factor(SUMO_foldChange_all$Type2 , levels =  unique(SUMO_genes$Type2))
```

```{r fig.height=10, fig.width=5.5}
# rearrange the experiments
SUMO_foldChange_all$exp <- factor(SUMO_foldChange_all$exp, levels =  c("1", "7", "3", "4", "5", "2", "6", "8", "9"))
p <- ggplot(data = SUMO_foldChange_all  , aes(x = exp, y = Symbol, fill = log2FoldChange)) +
  geom_tile()+
  theme_bw() +
  geom_raster() +
  geom_point(aes(size=sig)) +
     scale_size_manual(values=c(sig=1.5, no_sig=NA), guide="none") +
  facet_grid(Type2 ~. , scales = "free", space = "free") +
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, space = "Lab", 
   name="log2FC",
   limits = c(-2, 2), oob = scales::squish) +
  theme( axis.text.x = element_text(colour = "black"))

ggsave("SUMO_foldChange.png",plot = p, width = 6, height = 10, units = c("in"),
       dpi = 300)
```

# GO termn analysis
```{r}
library(gprofiler2)

getGO <- function(genelist = NULL, gene_bg = NULL){
GO_term <- gost(query = genelist, 
                organism = "athaliana", ordered_query = FALSE, 
                multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, 
                measure_underrepresentation = FALSE, evcodes = TRUE, 
                user_threshold = 0.05, correction_method = "fdr", 
                domain_scope = "annotated", custom_bg = gene_bg, 
                numeric_ns = "", sources = NULL)
GO_term$result
}

background_genes <- dataset_raw_dis0 %>% rownames_to_column(.) %>% select(rowname) %>% unlist() %>% as.vector()

GO_results_up <- getGO(genelist = DEG_gene_name_up, gene_bg = background_genes)
GO_results_down <- getGO(genelist = DEG_gene_name_down, gene_bg = background_genes)
```

# plot of GO term - combine all GOs and show by each experiment
```{r}
library(scales)
top10_GO_up <- GO_results_up %>% group_by(query)%>% dplyr::slice(1:10)

png(filename = "GO_top10_up.png", width = 13, height = 6, res = 300, units = "in")
ggplot(top10_GO_up, aes(x = intersection_size, y = term_name, fill = p_value))+
  geom_bar(stat="identity") +
  facet_grid(~query, scales = "free")
dev.off()

top10_GO_up$query <- str_remove(top10_GO_up$query, "exp")
top10_GO_up$query <- factor(top10_GO_up$query, levels = c("1", "7", "3", "4", "5", "2", "6", "8", "9"))

png(filename = "GO_top10_all_exp_up.png", width = 7, height = 8.5, res = 400, units = "in")
ggplot(data = top10_GO_up, aes(x = reorder(term_name, p_value),y=0 ,size = recall, color = -log(p_value))) +
  geom_point() +
  scale_color_gradient(low = "red", high = "blue") +
  theme_bw() + 
  xlab("") + 
  scale_x_discrete(labels = label_wrap(35)) +
  facet_grid(~query, scales = "free", space = "free") +
  coord_flip()+
  theme(axis.text = element_text(colour = "black"),
        axis.text.x = element_blank(),
        axis.title.x = element_blank()) 
dev.off()

#devtools::install_github("YuLab-SMU/ggtree")

library(ggtree)
top_GO_up_selective <- GO_results_up %>% group_by(query)%>% filter(str_detect(term_name, "salt|water|drought|abiotic|salicylic|osmotic"))

top_GO_up_selective$query <- str_remove(top_GO_up_selective$query, "exp")
top_GO_up_selective$query <- factor(top_GO_up_selective$query, levels = c("1", "7", "3", "4", "5", "2", "6", "8", "9"))


mat <- top_GO_up_selective %>% 
  select(query , p_value, term_name) %>%  
  pivot_wider(names_from = query, values_from = p_value) %>% 
  data.frame() %>% column_to_rownames(var = "term_name") %>% 
  replace(is.na(.), 0)


clust <- hclust(dist(mat %>% as.matrix())) # hclust with distance matrix

ddgram <- as.dendrogram(clust) # create dendrogram
plot(ddgram)

png(filename = "GO_top10-selective_all_exp_up_clusterTheTerms.png", width = 7, height = 8, res = 400, units = "in")
ggplot(data = top_GO_up_selective, aes(x = factor(term_name, levels = clust$labels[clust$order]),y=query ,size = recall, color = -log(p_value))) +
  geom_point() +
  scale_color_gradient(low = "red", high = "blue") +
  theme_bw() + 
  xlab("") + 
  scale_x_discrete(labels = label_wrap(45)) +
  coord_flip()+
  theme(axis.text = element_text(colour = "black"),
        axis.title.x = element_blank(),
        axis.text.x = element_text(colour = "black")) 
dev.off()




```

# Go - down-regulate
```{r}
top10_GO_down <- GO_results_down %>% group_by(query)%>% dplyr::slice(1:10)
top10_GO_down$query <- str_remove(top10_GO_down$query, "exp")
top10_GO_down$query <- factor(top10_GO_down$query, levels = c("1", "7", "3", "4", "5", "2", "6", "8", "9"))

mat_down <- top10_GO_down %>% 
  select(query , p_value, term_name) %>%  
  pivot_wider(names_from = query, values_from = p_value) %>% 
  data.frame() %>% column_to_rownames(var = "term_name") %>% 
  replace(is.na(.), 0)


clust_down <- hclust(dist(mat_down %>% as.matrix())) # hclust with distance matrix

ddgram_down <- as.dendrogram(clust_down) # create dendrogram
plot(ddgram_down)

png(filename = "GO_top10_all_exp_down_clusterTheTerms.png", width = 7, height = 9, res = 400, units = "in")
ggplot(data = top10_GO_down, aes(x = factor(term_name, levels = clust_down$labels[clust_down$order]),y=query ,size = recall, color = -log(p_value))) +
  geom_point() +
  scale_color_gradient(low = "red", high = "blue") +
  theme_bw() + 
  xlab("") + 
  scale_x_discrete(labels = label_wrap(40)) +
  coord_flip()+
  theme(axis.text = element_text(colour = "black"),
        axis.title.x = element_blank(),
        axis.text.x = element_text(colour = "black")) 
dev.off()
```



```{r}
module_df <- read.csv("gene_modules_GroupModel.txt")
all_module <- unique(module_df$colors)
experiment <- unique(only_salt_meta$Project)

# Initialize list to store all results
final_list <- list()

for (i in seq_along(all_module)) {
  module_name <- all_module[i]
  gene_in_module <- module_df %>% filter(colors == module_name)

  for (j in seq_along(experiment)) {
    exp_name <- experiment[j]
    deg_df <- DEG_list[[j]] %>% 
      rownames_to_column(var = "rowname") %>%
      filter(rowname %in% gene_in_module$gene_id) %>%
      select(rowname, log2FoldChange, padj) %>%
      mutate(module = module_name, exp = exp_name)

    final_list[[length(final_list) + 1]] <- deg_df
  }
}

# Combine all results into one data frame
final_result <- bind_rows(final_list)
final_result <- final_result %>% mutate(sig = if_else(padj <= 0.05, "sig", "ng"))

final_result_avg_fc <- final_result %>% select(log2FoldChange, module,exp) %>%
  group_by(module, exp) %>% 
  summarise(mean = mean(log2FoldChange)) %>% 
  mutate(exp2 = case_when(exp == "GSE275862" ~ "1",
                          exp == "GSE252395" ~ "2",
                          exp == "GSE270516" ~ "3",
                          exp == "GSE229217" ~ "4",
                          exp == "GSE124866" ~ "5",
                          exp == "PRJNA904764" ~ "6",
                          exp == "SUMOcode" ~ "7",
                          exp == "PRJNA904764_2" ~ "8",
                          exp == "PRJNA904764_3" ~ "9"))

final_result_avg_fc$exp2 <- factor(final_result_avg_fc$exp2, levels =  c("1", "7", "3", "4", "5", "2", "6", "8", "9"))
```

```{r}
png("log2fc_avg_gene_65modules.png", height = 10, width = 7, res = 300, units = "in")
ggplot(final_result_avg_fc, aes(exp, module, fill= mean)) + 
    geom_tile() +
   scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, space = "Lab", 
   name="log2FC",
   limits = c(-1, 1), oob = scales::squish) +
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90))
dev.off() 


#selected_modules in 20250407_GO_term_modules

selected_modules <- c("turquoise", "pink", "grey60", "green", "darkgrey", "brown")

png("log2fc_avg_gene_6-modules.png", height = 4, width = 5, res = 400, units = "in")
ggplot(final_result_avg_fc %>% filter(module %in% selected_modules), aes(exp2, module, fill= mean)) + 
    geom_tile() +
   scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, space = "Lab", 
   name="log2FC",
   limits = c(-1, 1), oob = scales::squish) +
  theme_bw()+
  theme(axis.text = element_text(colour = "black",size = 15),
        axis.title.x = element_blank())
dev.off() 

png("log2fc_avg_gene_7-modules.png", height = 5, width = 5, res = 300, units = "in")
ggplot(final_result_avg_fc %>% filter(module %in% selected_modules$query | module == "grey"), aes(exp, module, fill= mean)) + 
    geom_tile() +
   scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, space = "Lab", 
   name="log2FC",
   limits = c(-1, 1), oob = scales::squish) +
  theme_bw()+
  theme(axis.text.x = element_text(angle = 90))
dev.off() 

unique(selected_modules$query)
```

```{r}
turquoise_module_log2Fc <- final_result%>% filter(module == "turquoise") %>% select(rowname , log2FoldChange, exp) %>% spread(., key = "exp", value = "log2FoldChange")

write.csv(turquoise_module_log2Fc, "turquoise_module_log2Fc.csv", quote = F)
```

```{r}
arabidopsis_TF <- read.delim("Ath_TF_list.txt")
arabidopsis_TF$is_TF <- "yes"

#write.csv(arabidopsis_TF, "Ath_TF_list_mod.txt", quote = F)
```

```{r}
Turquoise_01_SUMO_1and2 <- read.csv("Turquoise_01_SUMO_1and2_OTS2.csv")

arabidopsis_TF_Turquoise_01_SUMO_1and2 <- arabidopsis_TF %>% filter(Gene_ID %in%unique(Turquoise_01_SUMO_1and2$g1, Turquoise_01_SUMO_1and2$g2))

arabidopsis_TF_freq <- as.data.frame(table(arabidopsis_TF$Family))

arabidopsis_TF_Turquoise_01_SUMO_1and2_percent <- data.frame(table(arabidopsis_TF_Turquoise_01_SUMO_1and2$Family)) %>% 
  left_join(arabidopsis_TF_freq, by = "Var1") %>% 
  mutate(percent = (Freq.x/Freq.y)*100)


# Reorder 'family' based on 'num1'
arabidopsis_TF_Turquoise_01_SUMO_1and2_percent$Var1 <- reorder(arabidopsis_TF_Turquoise_01_SUMO_1and2_percent$Var1, -(arabidopsis_TF_Turquoise_01_SUMO_1and2_percent$Freq.x))

png("TF_percent_in_turqouise_OTS2.png", height = 8, width = 12, units = "in", res = 300)
ggplot(arabidopsis_TF_Turquoise_01_SUMO_1and2_percent, aes(x = Var1, y = Freq.x, fill = "grey")) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = paste0(Freq.x, " (", round(percent, 1), "%)")), 
            hjust = -0.1, size = 3.5) +
  labs(
    x = "Gene Family",
    y = "Number of Genes"
  ) +
  theme_bw() +
  theme(legend.position = "none") +
  scale_y_continuous(limits = c(0, max(arabidopsis_TF_Turquoise_01_SUMO_1and2_percent$Freq.x)+3), expand = c(0, 0)) +
  coord_flip()
dev.off()

nrow(arabidopsis_TF_Turquoise_01_SUMO_1and2_percent)
```

```{r}

DEG_gene_name_up

# Count in how many experiments each gene appears
# Use unique() to avoid double-counting within a single experiment
gene_counts_up <- table(unlist(lapply(DEG_gene_name_up, unique)))

# Filter genes present in at least 2 experiments
Up_genes_in_2_or_more <- names(gene_counts_up[gene_counts_up >= 2])

#2268
write.csv(as.data.frame(Up_genes_in_2_or_more), "Arabidopsis_DEG_up-regulated_atleast_2exp.csv", quote = F)

# Down-regulated

gene_counts_down <- table(unlist(lapply(DEG_gene_name_down, unique)))

# Filter genes present in at least 2 experiments
Down_genes_in_2_or_more <- names(gene_counts_down[gene_counts_down >= 2])

#1594
write.csv(as.data.frame(Down_genes_in_2_or_more), "Arabidopsis_DEG_down-regulated_atleast_2exp.csv", quote = F)
```

# compare TF and gene related in salts
```{r}
arabidopsis_TF_Turquoise_01_SUMO_1and2

salt_genes_turquoise <- read.csv("turquoise_module_is_salt.csv")

TFandSalt <- intersect(arabidopsis_TF_Turquoise_01_SUMO_1and2$Gene_ID, salt_genes_turquoise$Gene.id)

arabidopsis_TF_Turquoise_01_SUMO_1and2 %>% filter(Gene_ID %in% TFandSalt) %>% select(Gene_ID )
```

